public/app/features/plugins/components/AppRootPage.tsx

```diff
@@ -19,112 +37,160 @@ interface Props extends GrafanaRouteComponentProps<RouteParams> {}
 
 interface State {
   loading: boolean;
-  portalNode: HtmlPortalNode;
   plugin?: AppPlugin | null;
-  nav?: NavModel;
+  pluginNav: NavModel | null;
 }
 
-export function getAppPluginPageError(meta: AppPluginMeta) {
-  if (!meta) {
-    return 'Unknown Plugin';
-  }
-  if (meta.type !== PluginType.app) {
-    return 'Plugin must be an app';
-  }
-  if (!meta.enabled) {
-    return 'Application Not Enabled';
-  }
-  return null;
-}
+const initialState: State = { loading: true, pluginNav: null, plugin: null };
 
-class AppRootPage extends Component<Props, State> {
-  constructor(props: Props) {
-    super(props);
-    this.state = {
-      loading: true,
-      portalNode: createHtmlPortalNode(),
-    };
-  }
+export function AppRootPage({ match, queryParams, location }: Props) {
+  const [state, dispatch] = useReducer(stateSlice.reducer, initialState);
+  const portalNode = useMemo(() => createHtmlPortalNode(), []);
+  const { plugin, loading, pluginNav } = state;
+  const sectionNav = useSelector(
+    createSelector(getNavIndex, (navIndex) => buildPluginSectionNav(location, pluginNav, navIndex))
+  );
+  const context = useMemo(() => buildPluginPageContext(sectionNav), [sectionNav]);
 
-  shouldComponentUpdate(nextProps: Props) {
-    return nextProps.location.pathname.startsWith('/a/');
-  }
+  useEffect(() => {
+    loadAppPlugin(match.params.pluginId, dispatch);
+  }, [match.params.pluginId]);
 
-  async loadPluginSettings() {
-    const { params } = this.props.match;
-    try {
-      const app = await getPluginSettings(params.pluginId).then((info) => {
-        const error = getAppPluginPageError(info);
-        if (error) {
-          appEvents.emit(AppEvents.alertError, [error]);
-          this.setState({ nav: getWarningNav(error) });
-          return null;
-        }
-        return importAppPlugin(info);
-      });
-      this.setState({ plugin: app, loading: false, nav: undefined });
-    } catch (err) {
-      this.setState({
-        plugin: null,
-        loading: false,
-        nav: process.env.NODE_ENV === 'development' ? getExceptionNav(err) : getNotFoundNav(),
-      });
-    }
-  }
+  const onNavChanged = useCallback(
+    (newPluginNav: NavModel) => dispatch(stateSlice.actions.changeNav(newPluginNav)),
+    []
+  );
 
-  componentDidMount() {
-    this.loadPluginSettings();
+  if (!plugin || match.params.pluginId !== plugin.meta.id) {
+    return <Page navId="apps">{loading && <PageLoader />}</Page>;
```

Comment:
If the feature toggle is off apps category does not exist

Reply:
does it just show `Page not found` in that case? how do we hit this scenario, i'm struggling to hit it locally 😅

Reply:
I'm pretty sure this `pageLoader` should be returned on first load of a plugin route (`!plugin`) or when switching directly between app plugins (`match.params.pluginId !== plugin.meta.id`). In my local testing though I cannot see it render.

Reply:
I think I need to fix this, it does flash "Page not found" for like a millisecond (I can see it using the Performance chrome tool that captures frames)